src/distilabel/steps/deita.py

```diff
@@ -0,0 +1,192 @@
+
+    _NearestNeighbors: Type["NearestNeighbors"] = PrivateAttr(...)
+
+    def load(self) -> None:
+        try:
+            from sklearn.neighbors import NearestNeighbors
+        except ImportError as ie:
+            raise ImportError(
+                "`scikit-learn` is not installed. Please install it using `pip install huggingface-hub`."
+            ) from ie
+
+        self._NearestNeighbors = NearestNeighbors
+
+    @property
+    def inputs(self) -> List[str]:
+        return ["evol_instruction_score", "evol_response_score", "embedding"]
+
+    @property
+    def outputs(self) -> List[str]:
+        return ["deita_score", "nearest_neighbor_distance"]
+
+    def _compute_deita_score(self, inputs: StepInput) -> StepInput:
+        """Computes the DEITA score for each instruction-response pair. The DEITA score is
+        the product of the instruction score and the response score.
+
+        Args:
+            inputs: The input data.
+
+        Returns:
+            The input data with the DEITA score computed.
+        """
+        for input_ in inputs:
+            evol_instruction_score = input_.get("evol_instruction_score")
+            evol_response_score = input_.get("evol_response_score")
+
+            if evol_instruction_score and evol_response_score:
+                deita_score = evol_instruction_score * evol_response_score
+            elif evol_instruction_score:
+                self._logger.warning(
+                    "Response score is missing for the instruction-response pair. Using"
+                    " instruction score as DEITA score."
+                )
+                deita_score = evol_instruction_score
+            elif evol_response_score:
+                self._logger.warning(
+                    "Instruction score is missing for the instruction-response pair. Using"
+                    " response score as DEITA score."
+                )
+                deita_score = evol_response_score
+            else:
+                self._logger.warning(
+                    "Instruction and response scores are missing for the instruction-response"
+                    " pair. Setting DEITA score to 0."
+                )
+                deita_score = 0
+
+            input_["deita_score"] = deita_score
+        return inputs
+
+    def _compute_nearest_neighbor(self, inputs: StepInput) -> StepInput:
+        """Computes the cosine distance between the embeddings of the instruction-response
+        pairs and the nearest neighbor.
+
+        Args:
+            inputs: The input data.
+
+        Returns:
+            The input data with the cosine distance computed.
+        """
+        embeddings = [input["embedding"] for input in inputs]
+        if self.normalize_embeddings:
+            embeddings = self._normalize_embeddings(embeddings)
+        self._logger.info("📏 Computing nearest neighbor distance...")
+        nn = self._NearestNeighbors(
+            n_neighbors=2, metric="cosine", algorithm="brute"
```

Comment:
is there a reason these are fixed?

Reply:
yes, we just want the distance with respect to the first nearest neighbor. The first neighbor will always be the record itself, therefore we take the second neighbor which is the real one. For the algorithm, we don't want any approximation and it's fast, therefore `brute`. metric can be a `RuntimeParameter`.